### DISCUSSION TITLE
Concern over the usage of adt.cfg

### DISCUSSION THREAD

**Discussion by marchbold** [airsdk/Adobe-Runtime-Support#1112] [Ideas]
I'd just like to raise a concern I have over the current development of the AIR SDK and the usage of placing project/user based configuration options in the AIR SDK.  I believe an SDK shouldn't be modified especially for project related configuration items eg `air.` prefix on package name or the path to a user installation (android sdk / ios sdk etc)

I've raised this several times in the past but feel I need to bring it up again.

I raised it with Adobe when they started using environment variables, particularly the `AIR_NOANDROIDFLAIR` which changed the output from `adt` for all Android applications. 

**I'm concerned over the usage of `adt.cfg` inside the AIR SDK to change the output for a project.** 

Currently if I change one of the `adt.cfg` options, it can break builds for other applications. This almost leads me to installing a version of the AIR SDK for every application I'm building, which makes no sense at all.

I understand that these options are ideally command line options, but given that we don't have access to how a lot of the IDEs use the command line directly I propose that we start using the application descriptor more extensively.

For me options like `AddAirToAppID`, `CreateAndroidAppBundle` and `KeepAndroidStudioOutput` make much more sense to be project specific and located in the `android` node of the application descriptor, particularly as they are directly related to the Android build, eg:
 
```xml
<application xmlns="http://ns.adobe.com/air/application/33.1">

    <!-- other elements -->

    <android>
        <manifestAdditions></manifestAdditions>
        <containsVideo>true</containsVideo>
        
        <addAirToAppID>false</addAirToAppID>
        <forceCreationAndroidAppBundle>false</forceCreationAndroidAppBundle>
        <keepAndroidStudioOutput>false</keepAndroidStudioOutput>
        
    </android>

...
```

I would love to hear other suggestions, but as the AIR SDK is moving forward again, (and quickly) I believe this is something we need to address and define a standard for creating options before we end up with more of a mess of configuration options.

**Comment by ajwfrost** [comment]
Thanks @marchbold and yes I think this does make sense (although perhaps with default options? so essentially like global settings in the SDK and then local overrides in the descriptor files? or do you think that's overkill..?)

My one (well, two) concerns are:
1) I'm not sure whether everyone actually edits the descriptor file. Tools like Animate and Flash Builder are potentially making changes to this themselves. In particular with the addition of placeholders e.g. for the name of the SWF file, plus some of the other Publish Settings etc that would mean the descriptor file gets modified by the tools anyway...
2) the descriptor file also gets copied as-is into the application.. so a minor point but this does mean that you're potentially exposing details about the development mechanism to the customers -> I for one try to avoid any machine-specific paths in any output (see for example the Flex SWC files and you will see the usernames and paths on the machines that they were built on....)

So I agree with this in principle, but would be interested in people's thoughts about whether a separate file would be a better location for this (e.g. "project settings", or - for example(!) - a `project.apm` file..?!)

I would love to have a graphical way to set all this up as well. Command-line options for sure but I think having a nice easy GUI that guides you through the options is a benefit (and it would need to be within our control, rather than the settings in the IDEs!).

One comment re. the central adt.cfg file: I agree that this isn't the best place for it (and similarly for the adt.lic file), since we've found (a) it means you have to keep copying/updating this every time you update the SDK, and (b) people put the SDK into locations that aren't writable!

So definitely a good idea to investigate this, thanks for pushing it... hoping that some other folk - particularly regular users of Animate, Flash Builder, IntelliJ, etc - will chip in.

**Reply by marchbold** [reply]
Definitely can have default values for all these in the sdk and these are overrides, however I don't believe they should be edited in the sdk by a user! The sdk should just be a download and install not something we have to edit. Configuration of the sdk should be done via other methods, I never edit other SDKs at all, unless I'm doing something highly abnormal. Probably the flags that output debug information for you to debug the SDK are an exception (eg `DebugOut` in adt.cfg). 

It's important to remember that as an app developer I when I update the AIR SDK I shouldn't have to then configure the SDK for my app, I should just be able to point my app at the new SDK and my project configuration should apply my application build settings.

1. Unless you are doing a very simple application you edit the application descriptor. Flash builder does automatically modify it for certain things (eg adding in some extension id's and setting some of the view options) and most IDEs auto insert the `content` swf (Animate an exception there), however everyone still edits it. Animate is probably one of the few avenues that requires you to switch a flag before you can edit it directly, but anyone who is doing a decent sized application would have done this.  

2. Definitely agree, I think that the descriptor is not the place for paths to the sdk's etc and please don't add those in there.  I think these would be better suited to a "user configuration" eg putting adt.cfg in the user's home directory or `~/.airsdk/adt.cfg`, and this is probably where the license file should be too?  You'll find most other frameworks do something similar (eg I have a few `.flutter*` files/dirs in the home directory and a `.android` directory where they store defaults and user configs). 

I just think it's important to separate these out from the SDK and have a clear definition of where they go. 

eg  
- sdk
  - internal debug settings? 
  - nothing commonly edited 
- user
  - paths that don't change often (defaults for android sdk, ios sdk, java location? )
  - license files 
- project 
  - everything else
  - output options, architectures, forcing targets 
  
One of the major points to consider here is the accessibility from existing IDEs. So adding in another project dependent config file isn't going to be easy, i.e. every IDE will have to add it to the adt command in some respect. Something like `apm`'s project configuration file `project.apm` is totally a great place for this project configuration  but I think it will be difficult to get the IDE builds to pick it up, when calling adt as they don't generally execute adt from the project directory.  This is the main reason I'm suggest the app descriptor as it's already plugged into the build process for all these IDEs.

But yeah really keen to hear other opinions on this, as it's been bugging me for a while.

**Reply by ajwfrost** [reply]
Thanks - all sounds good. I take your point re. the issue of adding another config element being more tricky for compatibility, so if we do have a way to set up path options etc in a user-specific folder then this part is sorted out and makes sense, and then anything that goes into the app descriptor file is just about describing the app and its settings..

Forcing the generation of particular output files (and I think we're only talking about Android here?) is an interesting one though. As the IDEs will often give 2 or 3 options (X86/ARM and maybe ARM 64-bit) but haven't yet caught up with X64 or App Bundles. So I'm wondering whether at the same time, we could simplify this so that we just have two options:
1) `GenerateMultipleAPKs` - if set, then we would create all 4 APK variants when a build is requested
2) `GenerateAppBundle` - if set, then we would create the App Bundle (as well as the requested APK?)
So for development/debugging, you can leave these off, just get a (32-bit) APK file and you can (I think?) install and run this on emulators and local devices. If you do need other APK variants for some reason (decreasingly likely) then you can switch that on and we'll just output all four of them with some filename suffixes. And then when you want to publish the bundle, you turn that on and then publish from the IDE and it will create a bundle (not sure whether this should be done instead of, or as well as, the file that's requested by the IDE....?)

I guess the question is to everyone else out there: what would be easiest and fit best with your tools and workflows!

**Comment by manishpatel0404** [comment]
Reading all of the above, I think you already have an answer.

Create a GUI for sdk settings, as @ajwfrost suggests and adopt the separation of concerns approach [sdk, user, project], as @marchbold suggests.

Being able to set up the user-specific folder is the starting point.

Apart from making the process of setting up the sdk easier, a GUI is the most feasible way to create a user-specific folder. The only other way is for users to create it themselves, which does not sound elegant.

1. The GUI executable/bat can be located in the root folder of the sdk, alongwith the readme and release notes.
2. The GUI can ideally have two tabs, one for "sdk settings" and another for "user settings".
3. The GUI can output the configuration file in the user-specific folder, which IDEs (over time) would learn to read and update.
4. For convenience, it can have a third tab which has links to the help/faq/support/tutorials/forums etc..
5. For growth and to attract new users, maybe yet another tab to a "Made with AIR" gallery, where developers can showcase their apps.
6. ...

**Reply by marchbold** [reply]
@manishpatel0404 What about the project specific settings, do you think the usage of the application descriptor is appropriate? Or have any other ideas? 

Probably this discussion is more about classifying these settings and setting a direction forward for adding more. A UI definitely would be useful in the future however I feel right now we just need to make sure we don't end up with a myriad of configuration options thrown all through the sdk.

**Reply by manishpatel0404** [reply]
Hi @marchbold - Sorry, I was not clear about the project settings. The project specific settings should continue to remain in the application descriptor, as you have mentioned in the example above. The user already has access to this file and is used to adding project-specific information in it. So, the overrides could be placed here.

**Comment by 2jfw** [comment]
Coming here (from recent post from today: https://forum.starling-framework.org/d/23340-air-config-settings) to give my two cents/requested feedback:
I also would welcome the usage of the application descriptor for these project related config settings. As already suggested, defaults in adt.cfg, overrides in app descriptor

**Comment by mrfrasier** [comment]
I haven't used Animate for a long time, but even when I did, I would manually edit certain parts of the manifest (Animate mangles it often). I have no issue with project-specific settings being overridable at the project level with a default at the SDK level.

Also OK with a separate file.

As for a GUI, I don't find that necessary, but get that many won't be comfortable. I'd say make it part of APM, but if you're using APM, you're probably already comfortable editing a config file. At the very least maybe APM can ask for some settings during project creation.

**Reply by marchbold** [reply]
I have definitely planned to make project configurations (and user configuration) settings part of `apm`. That is probably the main reason I've brought all this up, as I've started to investigate it all and it just seemed messed up to me. :)

**Comment by sjabberwocky** [comment]
It would be nice to have project specific settings inside descriptor or as some local `adt.cfg` file, which override settings from global `adt.cfg`.

For example, we have two projects - one with `air.` prefix and one without and changes in `adt.cfg` does not looks well.

**Reply by marchbold** [reply]
I totally agree on a local project file, but the issue I see with that is that it requires IDE changes, which was one of the reasons the adt.cfg in the sdk was created originally, ie. they couldn't get the IDE's to change the command line args easily so have applied global options to adt. This is why I proposed the app descriptor as it's already plugged into all these IDEs. 

Do you think it would be better to push the IDEs to update? 

I'm also concerned with adding another config file, (more clutter) when we already have something in place?

**Reply by sjabberwocky** [reply]
Personally I think that include settings from `adt.cfg` to descriptor is the best option.
For example, Google Play build should use `<forceCreationAndroidAppBundle>true</forceCreationAndroidAppBundle>` and Amazon build should use `<forceCreationAndroidAppBundle>false</forceCreationAndroidAppBundle>`.

At the same time `adt` may detect automatically if there is local `adt.cfg` near xml description without changes in command line so we do need to update all of IDEs. Especially for Flash Builder, which I still use, is hardly possible.

**Comment by blackjyn** [comment]
give my 2cents , we need  he **global/default adt.cfg** and the **local/project specific adt.cfg**.
The behavior is simple: project specific configs overrides the default one.

AND... why not also support **loading custom adt.cfg from specified location** like `adt -configfile=path/to/file.cfg`?

**Reply by marchbold** [reply]
Thanks for your thoughts, could you explain why you think we need a config inside the SDK?

**Reply by blackjyn** [reply]
simply for just making sure the default build configuration does exists

**Comment by marchbold** [comment]
@ajwfrost Your other post just reminded me about `mms.cfg`

Is that still relevant to AIR? What's stored in there?

**Reply by ajwfrost** [reply]
The `mms.cfg` file is still being read at start-up, and a fair number of the entries will still impact AIR although they're not so likely to be relevant I think...

See "Privacy and security settings" in:
https://www.adobe.com/content/dam/acom/en/devnet/flashplayer/articles/flash_player_admin_guide/pdf/latest/flash_player_32_0_admin_guide.pdf

There are a fair number of undocumented values too... my favourite is `AS3JIT` that can be 0 = interpret everything, 2 = JIT everything, 1 = mixed mode (normal). But most of those are just meant for debugging runtime issues rather than being settings that end users would change. The ideal would be that an application doesn't need anything set up there, and we adjust the gpu validation so that it actually works properly on Linux (i.e. it works out what capabilities the GPU has and therefore what Stage3D profiles are available, programmatically....)

thanks

**Reply by marchbold** [reply]
Interesting... For some reason I had thought it was a flash player thing only

**Comment by ajwfrost** [comment]
Hey everyone: can I please check for all you Android developers... would you have any need to have different build-tool versions being used for different applications? I'm wondering whether this should be a per-user setting, or a per-application one.

We've just noticed in the latest build tools, they have removed the dx.jar library which ADT would use for compiling the Java into Dex. Instead there is d8.jar which is the one that includes support for some of the Java 8 language features.

What we found a while back when d8.jar was first released was that some people couldn't compile their apps with it because they were using Java 1.7, in order to be able to support 1024-bit signing certificates still. (d8.jar cannot run in 1.7..) So in ADT, if you're creating an .apk file, if you're running Java 1.7 then we use dx.jar, if you're running Java 1.8+ then we use d8.jar, and if you're trying to use Java 8 features but are running in Java 1.7 then we fail with an error.

Anyway - the question is, would people need to maintain applications where they want to stick with Java 1.7 compatibility and use dx.jar from build tools 30.0.x, whilst also creating applications that should be using the latest build tools 31.0.x and beyond? So should this be a per-application setting?

In a similar vein .. any thoughts on whether we should allow people to configure the Gradle version and Gradle Android Plug-in version? There are such interdependencies on these (and maybe also on Android Studio versions) that it could get to be a bit of a minefield...


thanks

**Reply by marchbold** [reply]
We have been using 1.8 for a long time so I don't think this would ever affect us, however if you are looking to provide configuration options for this I'd suggest as java is generally a system level install that a user level configuration is appropriate here? Probably alongside the path to the android sdk etc? 

Are people still using 1.7 though (who can't upgrade to 1.8 at least)? Perhaps if Android is taking a lead here we should follow and deprecate usage of 1.7 for Android 31+?

**Reply by blackjyn** [reply]
Java 7 is dead. Go on with Java 8 or Java 11 when they are compatible . 
Java 8 is LTS which ended on March 2022. But it get extended support until December 2030:
https://www.oracle.com/java/technologies/java-se-support-roadmap.html

**Comment by ajwfrost** [comment]
Hi all

Feedback please! Below are our plans for how to update the application descriptor with platform-specific build/configuration options. My main concern is for folk using IDEs for Android development who want to create different APK and/or App Bundle outputs: how do we do this in a sensible way .. we're thinking that we give the option to generate multiple APKs with different architectures (though we may need to look at how this works with the version code requirements!) and then for building a bundle, we give the standard option like there is now of doing a bundle instead of an APK when an APK is requested, but also give an option to generate both APK and AAB outputs.

Would be grateful for any thoughts on these. Note that some of the settings would also be available as global setting (UncompressedExtensions, UseNativeCodesign, etc) but the app descriptor settings can override these.

thanks

Section | Name | Description
-- | -- | --
Windows | BuildArchitectures | Architecture(s) used when packaging a bundle or native target:   normally x86, can be set to x64 or both.
Android | AddAirToAppID | If set, overrides any global setting to add (or omit) the “air.” prefix   from an Android application ID
Android | BuildArchitectures | Set of architectures (armv7, armv8 etc) that will be created when the application is built, overriding any request from the IDE. Multiple   architectures will result in the APK files having “_arch” appended to the   filename (perhaps apart from the requested architecture, since the IDE may want to use this output file)
Android | CreateAppBundle | Whether to generate an app bundle build alongside or instead of an APK   despite the request by an IDE. App bundles will always contain all architecture   types.   Values can be “false” (same as omitted), “true” (just an app bundle is   created), or “both” (so that if an IDE requests an APK build, the APK is   built as well as the bundle)
Android | BuildLegacyAPK | When building an APK, if this is set then it will use the “old”   mechanism to generate the APK file (if false or omitted, it will use Gradle   similarly to how AAB is done)
Android | AndroidStudioProjectPath | Path for generating the Android Studio project. If omitted, the   project will be generated in a temporary folder and cleaned up afterwards. If   specified, the path will be cleaned up before generation and left in place   afterwards.
Android | UncompressedExtensions | To determine what resources should not be compressed within the APK   file
Android | JavaHome | JDK location to use for running third party tools
iOS | UseNativeCodesign | Whether to sign using internal tools or to use native macOS ‘codesign’   tool
iOS | SignSwiftFiles, OnlyIncludeSwiftUsedArchsInSupport, OnlyIncludeSwiftUsedArchsInPayload  | These settings relate to how Swift libraries are included in IPA   files. They should perhaps be available for configuration but not publicised,   as the default values should mean that the IPA files are then acceptable to Apple.
macOS | BundleCertificateName | For creating bundle/native types, this gives the name of a certificate   that should be used to sign the output application.
macOS | BuildArchitectures | Set of architectures to be included (x64, arm64) in case someone wants   an x64-only application that uses Rosetta2 on Apple silicon, or an app that   only works on Apple silicon. Default = both/universal.

**Reply by marchbold** [reply]
I think `JavaHome` shouldn't be here, but in a user config location eg `~/.airsdk/adt.cfg`.  I don't think it's a good idea to get user paths/setting in the app descriptor (user config is a good place for the android sdk path too if that's going to be required going forward)?

Also probably note on the `AndroidStudioProjectPath`, I assume that is going to be relative path to the app descriptor? 

Not really sure on the iOS swift signing options but I suppose you have some ideas about what you are exposing there? Ideally I hope we never have to use them :) I might say less is more here though? If these are SDK debugging options, might be better left in the SDK's adt.cfg? 

Otherwise looks pretty good, I'd suggest changing the names to be inline with other xml names though, i.e. starting with a lower case eg `addAirToAppID` 


Will these sit inside the platform options? eg:

```xml
<android>
    <build>
        <addAirToAppID>true</addAirToAppID>
        <androidStudioProjectPath>android/</androidStudioProjectPath>
    </build>
    <manifestAddtions> 

... 

</android>
```

**Reply by marchbold** [reply]
@ajwfrost Just checking is the latest sdk using `~/.airsdk/adt.cfg` ? What is the order or precedence with these? 

I will need to update apm to respect the "no air prefix" option appropriately so just checking first :)

**Reply by ajwfrost** [reply]
Hi

We had started moving settings across gradually, so in the current SDK it's using the `~/.airsdk/adt.cfg` file for information about JAVA_HOME and the Android SDK path. But, if you have an adt.cfg file in your SDK lib folder, then this would take precedence. So essentially we are looking at user-specified figures with the option of overriding them on a per-SDK basis in case that's needed.

Not sure if there's anything preventing us from moving the rest of the settings over in the same way, I will check with the team later this morning..

But I'm wondering whether there should be a way to get/set the configuration items via ADT, e.g. like you have with git config. That might remove the need for apm or anything to try to check where the setting may be stored or to update itself in the future if things change? So e.g. `adt -config set AddAirToAppID true`  or `adt -config get iOSPlatformSDK` etc. Whether we want to add an option here for setting/getting the values either globally (per user) or locally (per SDK) would be another option, although if the end-goal is to remove the per-SDK values then I suspect not and the 'get' would just follow the normal ADT logic for getting values, whatever that may end up being...

thanks

**Reply by marchbold** [reply]
Great stuff!

Just a note that seems a little backwards, would have expected it to go 

project (app xml)   >   user   >    global (sdk) 

in order of precedence.  Otherwise we are back in the situation I've mentioned above :)

**Reply by ajwfrost** [reply]
The rationale for having per-SDK settings override per-user settings was that a user may have multiple SDKs installed and want different behaviour from different ones. But we're not exactly "real world" users here (!) so if you think it's more likely that people will want to have one set of common settings, then we can go with per-user being the main repository, with per-project (app descriptor) values overriding these.

It kind of means that I don't see much point in having an adt.cfg file within the SDK location ... I was always finding it a pain to copy the values from one SDK to another, so maybe we should scrap that option altogether. For the interim we can do what we have with the license file, i.e. if we find it in the user folder then we use it; if not, but if it's in the SDK folder, then we copy that to the user folder. Then we are essentially just using the user folder. So for configuration, we could check there is a settings file in the user folder: if there isn't, we'll just copy the one from the SDK folder into it, and then everything starts using that.

Would that work? It would mean we just have the two levels:
- project-specific, set via the application descriptor, for elements that make sense to be per-project
- user-specific, set via ~/.airsdk/adt.cfg, for things like tool/SDK paths and common elements, and for defaults where there is no project-specific override

This may keep it simpler...?

**Reply by marchbold** [reply]
I wouldn't suggest copying it, just use it if it is there or use what's in the sdk otherwise. It should waterfall for each value, i,e the user adt.cfg may only specify some of the values. 

I would expect values in the sdk to be useful still 

eg when debugging a version of the sdk and i want to enable log outputs. I don't want to do this at the user level, just fora particular sdk version that i'm trying to debug to help send info to you. Similarly for the "keep studio project" etc. 

I suppose I see the values that are used for debugging the sdk still valuable there, but ones that affect project outputs and sdk locations are the ones most likely to be set at a project / user level.

**Reply by Klug76** [reply]
Hi!
Is there a way to build just project for android studio without building apk/aab? This would give a significant increase in build speed if a pre-build step was required (for example, adding a huawei plugin to settings.gradle).

**Reply by ajwfrost** [reply]
@Klug76 that's a good idea ... currently no, but if we added a new 'target' rather than APK or AAB we could use the same packager and just omit the Gradle step at the end... I think this should be pulled out into a new request -> #1818
